codalab/lib/bundle_cli.py

```diff
@@ -501,43 +504,58 @@ def simple_user_str(user):
 
-    @staticmethod
-    def parse_target(client, worksheet_uuid, target_spec):
-        """
-        Helper: A target_spec is a [worksheet_spec//]bundle_spec[/subpath].
-        Returns: (bundle_uuid, subpath) where the bundle_uuid is the uuid of the
-        bundle matching the bundle spec, from the worksheet matching the given
-        worksheet spec if one is provided
-        """
-        worksheet_spec = None
-        if '//' in target_spec:
-            worksheet_spec, target_spec = target_spec.split('//', 1)
-        if os.sep in target_spec:
-            bundle_spec, subpath = tuple(target_spec.split(os.sep, 1))
+    def resolve_target(self, default_client, default_worksheet_uuid, target_spec, allow_remote=True):
+        """
+        Input: Target spec in the form of
+        [<instance>::][<worksheet_spec>//]<bundle_spec>[/<subpath>]
+            where <bundle_spec> is required and the rest are optional.
+
+        Returns:
+            - client: A client connected to the instance the target is from if allow_remote is True and an instance is specified,
+                otherwise default_client
+            - worksheet_uuid: The uuid of the worksheet the target bundle is from,
+                a new uuid if the target spec includes a worksheet spec
+                same as default_worksheet_uuid otherwise
+            - bundle_uuid: The uuid of the target bundle
+            - subpath: The subpath from the target_spec
+        Raises UsageError if allow_remote is False but an instance is specified in the target_spec
+        """
+        instance, worksheet_spec, bundle_spec, subpath = parse_target_spec(target_spec)
+
+        if instance is not None:
+            if not allow_remote:
+                raise UsageError('Cannot execute command on a target on a remote instance. Please remove the instance reference (i.e. "prod::" in prod::worksheet//bundle)')
+            aliases = self.manager.config['aliases']
+            if instance in aliases:
+                instance = aliases.get(instance)
+            client = self.manager.client(instance)
         else:
-            bundle_spec, subpath = target_spec, ''
-
-        if worksheet_spec:
+            client = default_client
+        if worksheet_spec is not None:
             worksheet_uuid = BundleCLI.resolve_worksheet_uuid(client, '', worksheet_spec)
+        else:
+            worksheet_uuid = default_worksheet_uuid
 
         # Resolve the bundle_spec to a particular bundle_uuid.
         bundle_uuid = BundleCLI.resolve_bundle_uuid(client, worksheet_uuid, bundle_spec)
-        return (bundle_uuid, subpath)
 
-    def parse_key_targets(self, client, worksheet_uuid, target_specs):
+        return (client, worksheet_uuid, bundle_uuid, subpath)
+
+    def resolve_key_targets(self, client, worksheet_uuid, target_specs):
         """
         Helper: target_specs is a list of strings which are [<key>]:<target>
+        Returns: List<Tuple<key, Tuple<bundle_uuid, subpath>>>
```

Comment:
I'd just write `[(key, (bundle_uuid, subpath)), ...]` to make it easier to read.